src/lib/cost_estimation/abstract_cost_estimator.cpp

```diff
@@ -0,0 +1,105 @@
+#include "abstract_cost_estimator.hpp"
+
+#include <queue>
+#include <unordered_set>
+
+#include "logical_query_plan/abstract_lqp_node.hpp"
+#include "logical_query_plan/lqp_utils.hpp"
+#include "statistics/abstract_cardinality_estimator.hpp"
+
+namespace opossum {
+
+AbstractCostEstimator::AbstractCostEstimator(const std::shared_ptr<AbstractCardinalityEstimator>& cardinality_estimator)
+    : cardinality_estimator(cardinality_estimator) {}
+
+Cost AbstractCostEstimator::estimate_plan_cost(const std::shared_ptr<AbstractLQPNode>& lqp) const {
+  // Sum up the costs of all operators in the plan, while making sure to cost each operator exactly once, even in the
+  // presence of diamond shapes.
+
+  // Breadth-first iteration of plan
+  auto bfs_queue = std::queue<std::shared_ptr<AbstractLQPNode>>{};
+  auto visited = std::unordered_set<std::shared_ptr<AbstractLQPNode>>{};
+
+  bfs_queue.emplace(lqp);
+
+  auto cost = Cost{0};
+
+  while (!bfs_queue.empty()) {
+    const auto current_node = bfs_queue.front();
+    bfs_queue.pop();
+
+    const auto visited_iter = visited.find(current_node);
+    if (visited_iter != visited.end()) {
+      continue;
+    }
+    visited.emplace(current_node);
+
+    const auto cached_cost = _get_subplan_cost_from_cache(current_node, visited);
```

Comment:
I hope this explains it somewhat:

```c++
    // Not every visited node has a cached cost. Costs are only cached for the roots of the plans for which 
    // `AbstractCostEstimator::estimate_plan_cost()` is called. For subplans below the root, costs cannot be cached
    // because their costs are never aggregated due to the traversal being a BFS.
```